src/python/grpcio_tests/tests/unit/_tcp_proxy.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2019 the gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Proxies a TCP connection between a single client-server pair.
+
+This proxy is not suitable for production, but should work well for cases in
+which a test needs to spy on the bytes put on the wire between a server and
+a client.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import datetime
+import select
+import socket
+import threading
+
+_TCP_PROXY_BUFFER_SIZE = 1024
+_TCP_PROXY_TIMEOUT = datetime.timedelta(milliseconds=500)
+
+
+def _init_listen_socket(bind_address):
+    listen_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    listen_socket.bind((bind_address, 0))
+    listen_socket.listen(1)
+    return listen_socket, listen_socket.getsockname()[1]
+
+
+def _init_proxy_socket(gateway_address, gateway_port):
+    proxy_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    proxy_socket.connect((gateway_address, gateway_port))
+    return proxy_socket
+
+
+class TcpProxy(object):
+    """Proxies a TCP connection between one client and one server."""
+
+    def __init__(self, bind_address, gateway_address, gateway_port):
+        self._bind_address = bind_address
+        self._gateway_address = gateway_address
+        self._gateway_port = gateway_port
+
+        self._byte_count_lock = threading.RLock()
```

Comment:
optional: In your code here, the modification to `*_byte_count` always happen in the same thread. That means there is no race in writing, so the modification will be correct. Though the read of the `*_byte_count` is happen in another thread, the GIL guarantees the stored value won't be corrupted. So, I presume remove this lock, your code is still correct.

Reply:
`reset_byte_count` isn't guaranteed to be run on the same thread.